neuralpp/symbolic/lazy_normalizer.py

```diff
@@ -5,33 +5,40 @@
 from .parameters import sympy_evaluate
 import neuralpp.symbolic.functions as functions
 from .general_normalizer import GeneralNormalizer
+from .profiler import Profiler
 
 
-def _normalize(expression: Expression, context: Z3SolverExpression) -> Expression:
-    match expression:
-        case Constant() | Variable():
-            return expression
-        case FunctionApplication(function=Constant(value=functions.conditional),
-                                 arguments=[condition, then, else_]):
-            if context.is_known_to_imply(condition):
-                return _normalize(then, context)
-            elif context.is_known_to_imply(~condition):
-                return _normalize(else_, context)
-            else:
-                return if_then_else(condition,
-                                    _normalize(then, context & condition),
-                                    _normalize(else_, context & ~condition))
-        case FunctionApplication(function=function, arguments=arguments):
-            arguments = [_normalize(argument, context) for argument in arguments]
-            return expression.new_function_application(function, arguments)
-        case QuantifierExpression():
-            return GeneralNormalizer.normalize(expression, context)
-        case _:
-            raise ValueError(f"invalid expression {expression}")
+class LazyNormalizer(Normalizer):
+    def __init__(self, profiler: Profiler = None):
+        if profiler is None:
+            self.profiler = Profiler()
+        else:
+            self.profiler = profiler
+        self._general_normalizer = GeneralNormalizer(self.profiler)
 
+    def normalize(self, expression: Expression, context: Z3SolverExpression) -> Expression:
+        with self.profiler.profile_section("normalization"):
```

Comment:
Do we not need `sympy_evaluate(True)` anymore? If so, then we can probably delete the import at the top of the file

Reply:
yes